src/decision/src/VisionDecision.cpp

```diff
@@ -24,13 +30,321 @@ VisionDecision::VisionDecision(int argc, char **argv, std::string node_name) {
+    // the higher the desired angle, the higher the angular speed
+    return mapRange(speedToMap, 0, 90, 0, 100);
+
+}
+
+/**
+ *  Returns the desired forward speed
+ *
+ *  @param desiredAngle
+ *      angle robot will turn
+ *  @returns double
+ *      moving speed percentage of robot
+ */
+double VisionDecision::getDesiredSpeed(double desiredAngle){
+    double speedToMap = abs((int)desiredAngle);
+    // the higher the desired angle the lower the linear speed.
+    return 100 - mapRange(speedToMap, 0, 90, 0, 100);
+}
+
+
+/* Helper functions for functions that determine robot movement. */
+
+/**
+ * Determines the middle of the white line given a row
+ *
+ * @param startingPos which column to start parsing in
+ * @param row the row to parse
+ * @param rightSide determines whether to parse from the right or the left
+ * @param image_scan the image to parse
+ */
+int VisionDecision::getMiddle(int startingPos, int row, bool rightSide, const sensor_msgs::Image::ConstPtr& image_scan){
+
+    int incrementer;
+    int startPixel, endPixel;
+
+    // Depending on chosen side, determine where to start
+    // and how to iterate.
+    if(rightSide){
+        incrementer = -1;
+    }else{
+        incrementer = 1;
+    }
+
+    // Find first pixel of the white line in a certain row.
+    startPixel = VisionDecision::getStartPixel(startingPos, incrementer, row, image_scan);
+
+    // Find last pixel of the white line in a certain row.
+    endPixel = VisionDecision::getEndPixel(startPixel, incrementer, row, image_scan);
+
+    // Return average of the two pixels.
+    return (startPixel + endPixel) / 2;
+}
+
+/**
+ * Returns the white pixel right after the black space in between
+ * the line and the left side or right side of the screen.
+ *
+ * @param startingPos column to start parsing
+ * @param NOISEMAX how large noise can be
+ * @param incrementer decides whether to parse from the left or from the right
+ * @param row determines the row to parse
+ * @param image_scan the image to parse
+ *
+ * @returns the white pixel's column position, null if none found
+ */
+int VisionDecision::getStartPixel(int startingPos, int incrementer, int row,
+                     const sensor_msgs::Image::ConstPtr& image_scan){
+    int column = startingPos;
+    int whiteVerificationCount = 0;
+    int blackVerificationCount = 0;
+    int startPixel = null;
+    int toBeChecked = null;
+
+    // Find starting pixel
+    while(column < image_scan->width && column >= 0){
+        // If white pixel found start verifying if proper start.
+        if(image_scan->data[row*image_scan->width + column] != 0){
+            blackVerificationCount = 0;
+            // This pixel is what we are checking
+            if (toBeChecked == null)
+                toBeChecked = column;
+
+            // Determine whether toBeChecked is noise
+            whiteVerificationCount++;
+            if (whiteVerificationCount == NOISEMAX && startPixel == null)
```

Comment:
May want to change `NOISEMAX` to be relative to the size of the image. (ie. instead of 30 pixels, 30 pixels for every 1000 pixels in the image width)